BUG: Improve numerical precision/convergence failures of smirnov/kolmogorov

pvanmulbregt opened the pull request:
Addresses #7426, #7455, #7456, #7492.
scipy.special.cephes:
A rewrite of the algorithms in the Kolomgorov-Smirnov statistics package.
Kolmogorov: Improve efficiency of kolmogorov().
Fixed NANs, false convergence, unbounded computation in kolmogi().
Added dedicated functions (kolmogc/kolmogci/ kolmogp) to compute
the CDF, PPF and PDF.

Smirnov: Improve accuracy of smirnov().  Computation suffered from underflow
and/or denormalization, resulting in probabilities occasionally off by orders
of magnitude.  The SF/CDF is now usually with 1 ULP, the PDF within a few ULP.
Fixed NANs, convergence failures, false convergence in smirnovi().
Added dedicated functions (smirnovc/smirnovci/smirnovp) to compute
the CDF, PPF and PDF.
Added additional testing of K-S, removed xfail on previous tests.

scipy.stats: ksone and kstwobign classes.
Added dedicated _sf, _pdf, _isf methods to call new K-S functionality.
Updated _cdf and _ppf methods to make appropriate calls.

The accuracy improvements to smirnov() required increased precision
in some parts of the computation. Uses the recent port of the dd_real part of
the qd package.

Allow FuncData testing to specify a list/array of dtypes to apply.
FuncData.check(dtypes=[int, float, float]) will treat the first 3 columns
as type int, float, float respectively, rather than blanketly
treating all as the underlying type of the data.  Needed for testing
some C functions that take an int as a parameter.

pvanmulbregt commented:
This PR is now stable for review (code build/passes Py2/Py3, Windows/Linux; doc and examples build under Py2/Py3).

The PR adds 6 functions to `scipy.special`, implemented in C in `scipy/special/cephes/kolmogorov.c`: 
- `smirnovc`, `smirnovci`, `smirnovp` (joining already existing `smirnov`, `smirnovi`).
- `kolmogc`, `kolmogci`, `kolmogp` (joining already existing `kolmogorov`, `kolmogi`)

The function names take the form  `('kolmog'|'smirnov')+suffix`.  The suffixes are: no suffix or `''` (computes the Survival Function SF), `'c'` (computes the CDF), `'ci'` (inverse CDF or PPF), `'i'` (inverse Survival Function ISF), `'p'` (derivative, or prime, of the SF). The already existing `kolmogorov` is the only exception to this rule.
The PPF/ISF functions for both Smirnov and Kolmogorov now use a bracketed Newton-Raphson root-solver with legal initial point and a tight bracket, and use an exactly computed derivative.
The Smirnov CDF/SF/PDF functions use some double-double arithmetic to avoid accuracy loss when computing large powers, and take care to avoid underflow/overflow/denormalization in intermediate results.
The Kolmogorov CDF/SF/PDF functions use an identity for Jacobi theta functions to replace a high condition number sum of a few hundred terms with a sum of at most 4 positive terms.
Examples have been added to the doc as per gh-7168.
Tests have been updated or added, and `pytest.mark.xfail` decorators removed.
The `_cdf/_isf/_ppf/_isf/_pdf` methods of the probability distributions `scipy.stats.ksone` and `scipy.stats.kstwobign` have been updated to call the new K-S functions.

pvanmulbregt commented on scipy/special/_cephes.pxd:
```diff
@@ -93,9 +93,15 @@ cdef extern from "cephes.h" nogil:
     double kolmogi(double p)
+    double kolmogc(double x)
+    double kolmogci(double p)
+    double kolmogP(double x)
```
`kolmogP` should be `kolmogp`

pvanmulbregt replied:
Fixed.

chrisb83 commented:
Is someone already reviewing this? since i am interested to understand the details of the KS test, I was thinking to look into the arXiv articles anyways and could then give feedback on the implementation as well.

pvanmulbregt commented:
I'm not aware of anyone, so I'd be happy for you take a look.

chrisb83 commented:
I had a look the arxiv paper https://arxiv.org/abs/1803.00426 which treats the asymptotic distribution of the one-sided one-sample test statistic. 

I guess the main quantity of interest is the survival function / CDF and the proposal to compute it more accurately looks very good, the approach to use an identity of the Theta function is also stated in Marsaglia et al "Evaluating Kolmogorov's Distribution". This article also gives a C algorithm to compute the CDF of the exact statistic `D_n` instead of the asymptotic Distribution. This might be something interesting for a follow-up project, `R` uses this implementation (https://www.rdocumentation.org/packages/dgof/versions/1.2/topics/ks.test)

In the `R`documentation, I liked the idea to warn if there are ties since the test should only be applied to continuous distributions. Is that implemented in scipy as well?

I will look at the implementation soon as well as the one-sided case (btw, I think both arXiv papers state that `P[D_n > x] = 2 P[D_n^+ > x]` and this is not true, it should be <=, but this is not relevant  for the rest of the article).

pvanmulbregt commented:
Thanks Chris.  Following up on your comments 
1. The paper https://arxiv.org/abs/1803.00426 actually treats the asymptotic distribution of the _**two**_-sided one-sample test statistic.  [The other paper https://arxiv.org/abs/1802.06966 deals with one-sided test statistics.]
2. Indeed Marsaglia provided an algorithm for computing the exact distribution of `D_n`.  Carvalho ["An Improved Evaluation of Kolmogorov's Distribution"](https://www.jstatsoft.org/v065/c03) gave a different algorithm.  I have implementations of both.  They have different memory/computation tradeoffs. 
3. The code to warn if there are ties would be in the function`stats/stats.py:kstest()`.  Currently no such warning exists.  I've actually seen ties when applying to continuous data, due to clipping or other reasons. 
4. As you note `P[D_n > x] = 2 P[D_n^+ > x]` is not true in general, as `P[D_n > x] = 2 P[D_n^+ > x] - P[(D_n^+ > x) and (D_n^- > x)]`. But the equality is true if `x>= 0.5` as the intersection is then empty, and I hope that I specified that condition in the papers!

chrisb83 commented on scipy/special/add_newdocs.py:
```diff
@@ -4407,22 +4407,218 @@ def add_newdoc(place, name, doc):
+
+    Returns
+    -------
+    float
```
Maybe write y : float and repeat that kolmogorov(y) = p here again for clarity? y is array_like as p I guess, same for the other functions below.

chrisb83 commented on scipy/stats/_continuous_distns.py:
```diff
@@ -40,11 +40,20 @@ class ksone_gen(rv_continuous):
     %(default)s
 
     """
+    def _pdf(self, x, n):
```
it would be nice to expand the description of ksone_gen and kstwobign (potentially reuse text from special/add_newdocs.py ), maybe also reference to the implemention of the functions in scipy.special

chrisb83 commented on scipy/special/add_newdocs.py:
```diff
@@ -4407,22 +4407,218 @@ def add_newdoc(place, name, doc):
+    >>> kolmogorov([0, 0.5, 1.0])
+    array([ 1.        ,  0.96394524,  0.26999967])
+
+    Compare a sample of size 1000 drawn from a Laplace(0, 1) distribution against
```
At first glance, I found this example confusing. I guess it is an illustration that the KS test rejects the hypothesis that the sample x is drawn from a normal distribution by computing the test statistic and then computing the p-value. I suggest to add a few more comments, maybe one could even add a line with kstest(x, "norm") which gives KstestResult(statistic=0.05828560005373079, pvalue=0.0021467598952917832) to confirm the result.

chrisb83 commented:
Hi,

regarding 3: ok, i saw now that you did not touch the KS test itself as it will simply use your new implementation via `stats.ksone`. Even though ties might appear due to clipping / rounding, I would still consider a warning useful since it raises awareness of potential issues. But it is not importnt in this context here about improving the precision of the functions.

regarding 4: indeed, the condition `x > 0.5` is stated in the paper, I did not read it carefully.

I have had a look the paper on the one-sided distribution, Overall it looks much simpler in that case, with the exception of your analysis on the precision which I did not go thorough. I will try to go through the code soon when i find time.

chrisb83 commented on scipy/special/add_newdocs.py:
```diff
@@ -4407,22 +4407,218 @@ def add_newdoc(place, name, doc):
+    Cumulative distribution function of Kolmogorov distribution
+
+    Returns the cumulative distribution function of
+    Kolmogorov's limiting distribution, ``D_n*\sqrt(n)`` for large n, of a
```
Suggest to write 'as n goes to infinity' instead of large n
sqrt instead of \sqrt since not in math mode

chrisb83 commented on scipy/special/add_newdocs.py:
```diff
@@ -6348,16 +6544,206 @@ def add_newdoc(place, name, doc):
+    >>> smirnov(5, [0, 0.5, 1.0])
+    array([ 1.   ,  0.056,  0.   ])
+
+    Compare a sample of size 5 drawn from a Normal(0.5, 1) distribution against
```
I would say norm(0, 1)

chrisb83 commented on scipy/special/add_newdocs.py:
```diff
@@ -6348,16 +6544,206 @@ def add_newdoc(place, name, doc):
+    Returns
+    -------
+    float
+        The value(s) of smirnov(n, d), Prob(Dn >= d)
```
Isn't that Dn+ / Dn- ?

chrisb83 commented on scipy/special/add_newdocs.py:
```diff
@@ -6348,16 +6544,206 @@ def add_newdoc(place, name, doc):
+    Returns
+    -------
+    float
+        The value(s) of smirnovc(n, d), Prob(Dn <= d)
```
Again Dn+/-?

chrisb83 commented on scipy/special/add_newdocs.py:
```diff
@@ -6348,16 +6544,206 @@ def add_newdoc(place, name, doc):
+
+    See Also
+    --------
+    smirnov  : The SF for the distribution
```
Write out survival function as above

chrisb83 commented on scipy/special/add_newdocs.py:
```diff
@@ -6348,16 +6544,206 @@ def add_newdoc(place, name, doc):
+
+    Inverse to `smirnovc`
+
+    Returns `d` such that ``smirnovc(n, d) = p``.
```
== p

chrisb83 commented on scipy/special/add_newdocs.py:
```diff
@@ -6348,16 +6544,206 @@ def add_newdoc(place, name, doc):
+
+    Derivative of `smirnov`.
+
+    As ``smirnov(n, d)`` is the Survivor Function, ``smirnovp(n, d)`` is the negative of
```
Survival

chrisb83 commented on scipy/special/add_newdocs.py:
```diff
@@ -6348,16 +6544,206 @@ def add_newdoc(place, name, doc):
+    >>> cdfs = target.cdf(x)
+    >>> cdfs
+    array([ 0.41704346,  0.76327829,  0.82936059,  0.92589857,  0.99899518])
+    # Construct the Empirical CDF and the K-S statistics (Dn+, Dn-, Dn)
```
Do you compute Dn? Could even be useful to add to the example and compute asymptotic p-value to show the link to the kolmog functions. Though n=5 not ideal for asymptotic result :)

chrisb83 commented on scipy/special/add_newdocs.py:
```diff
@@ -6348,16 +6544,206 @@ def add_newdoc(place, name, doc):
     Inverse to `smirnov`
 
-    Returns ``e`` such that ``smirnov(n, e) = y``.
+    Returns `d` such that ``smirnov(n, d) = p``, the critical value.
```
== p, remove . at the end

chrisb83 commented:
Sorry, it took me some time to get back to you. I made some minor suggestions to the doc strings of the smirnov functions.

One additional remark: the references are currently hidden in the .c file, right? I would make them more visible.

pvanmulbregt commented:
Modified most documentation changes identified in the review.
Changed `Normal(0, 1)` to `N(0, 1)` rather than `norm(0, 1)`, as I want to emphasize the theoretical probability distribution, not the SciPy approximation to it.
I left the references  in the c file. as they pertain to the particular implementation(s) there, rather than being informational on the functions in general.

rgommers commented:
> Is Travis functioning correctly? I'm seeing timeout errors after 10min with no output in one build and errors in completely independent parts of the code in another build, and the others all pass.

That's not uncommon. All CI systems have that to some extent. If you're logged in for TravisCI with your GitHub credentials, you can restart just that build - I've just done that.

person142 commented:
So I am unfortunately not going to have time to review this carefully, but something I wanted to throw out: do we need to add so many functions to the public API, or should we leave them private and just use them in `stats`? There are already *way* too many public functions in `special`.

person142 commented on scipy/special/_legacy.pxd:
```diff
@@ -135,8 +135,26 @@ cdef inline double smirnov_unsafe(double n, double e) nogil:
     _legacy_cast_check("smirnov", n, 0)
     return smirnov(<int>n, e)
 
+cdef inline double smirnovc_unsafe(double n, double e) nogil:
```
As we discussed before, we don't want to add more unsafe versions of functions.

pvanmulbregt replied:
Yes, I tried not adding the additional entries to `_legacy.pxd` but immediately incurred errors when testing `stats`.  These seem related to the signature and disappeared when I added the entries into `_legacy.pxd`.  E.g.
```
$ python runtests.py -n -j4 -s stats
...
self = <scipy.stats._continuous_distns.ksone_gen object at 0x10cd616a0>, q = array([0.5]), n = array([1.])

    def _ppf(self, q, n):
>       return sc.smirnovci(n, q)
E       TypeError: ufunc 'smirnovci' not supported for the input types, and the inputs could not be safely coerced to any supported types according to the casting rule ''safe''

n          = array([1.])
q          = array([0.5])
self       = <scipy.stats._continuous_distns.ksone_gen object at 0x10cd616a0>

../../scipy/stats/_continuous_distns.py:53: TypeError
```

chrisb83 commented:
> So I am unfortunately not going to have time to review this carefully, but something I wanted to throw out: do we need to add so many functions to the public API, or should we leave them private and just use them in stats?

I think this is a good suggestion. The survival functions are well known where the other functions are probably only used together with the distribution in `stats`.

chrisb83 commented:
> Modified most documentation changes identified in the review.

Thanks, I think the documentation looks good now. 

At some point, one should improve the description of the distributions `ksone` and `kstwobign`, potentially including the references there. I always find it useful to know where i can find more information.

I had a look at testing, it looks comprehensive in my view. I did not check the C implementation in more detail (i only have very basic C knowledge anyways). So I will stop my review here.

I think this is a big improvement and I hope it can be merged soon.

rgommers commented:
> Doing this presumably would mean that any functions calling the C-code directly would break and need to change. (I looked and didn't find any of those inside SciPy itself. Is that sufficient?).

We don't support calling the C code directly, so that's not a problem.

> Any code which accessed smirnov() via scipy.special.cython_special.smirnov() would also no longer work as smirnov itself is no longer a cython function. Is that direct method of access supported?

Yes, `cython_special.smirnov` is part of the public API.

> Q: Is this approach of adding options to existing functions preferable to adding functions?

If it can't be done without removing from the Cython interface, then I think the answer is no. @person142 or @pv may have more elaborate suggestions?

person142 commented:
My more elaborate suggestion would be supporting optional arguments in the Cython wrappers. I have been meaning to do this for a while, but you know, time. In the meantime, is it so bad for the functions to be private? I doubt they would be used outside the stats context.

pvanmulbregt commented:
@person142  How does one make a C function inside scipy/special/cephes/kolmogorov.c private, but still accessible by Python inside scipy/stats/_continuous_distns.py?

person142 commented:
Do everything in the normal way, but prefix it with an underscore in `functions.json`. Then import it explicitly from `special._ufuncs` in stats when you need it. See e.g. `_sinpi` for an example; i.e. this works fine:

```
from scipy.special._ufuncs import _sinpi
```

pvanmulbregt commented:
No longer add  `kolmogc/kolmogci/kolmogp/smirnovc/smirnovci/smirnovp` functions to the public `special` API, but keep them private and import them directly into `scipy/stats/_continuous_distns.py` for use in `stats.{ksone,kstwobnign}`. Added a `Notes` section to the docstrings of the existing `smirnov/smirnovi/kolmogorov/kolmogi` functions to suggest that directly using the two K-S distributions in `stats` is the more accurate way of retrieving the CDF/PPF and the PDF.

pvanmulbregt commented:
Tests pass except for an `mpmath.libmp.BACKEND=="gmpy"` assertion on Travis-ci running OSX.  Presumably this is covered by gh-9106 and gh-9127. Just ignore the failure?

andyfaff commented:
I'll merge the Travis modification tomorrow which removes the assert for
gmpy on OSX. Might be worth waiting until that's merged?

On Sun., 12 Aug. 2018, 11:05 Paul van Mulbregt, <notifications@github.com>
wrote:

> Tests pass except for an mpmath.libmp.BACKEND=="gmpy" assertion on
> Travis-ci running OSX. Presumably this is covered by gh-9106
> <https://github.com/scipy/scipy/issues/9106> and gh-9127
> <https://github.com/scipy/scipy/pull/9127>. Just ignore the failure?
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scipy/scipy/pull/8737#issuecomment-412311419>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAq51jbnoHbV53hhxfHwa_pQeFmPsecyks5uP39SgaJpZM4TVxN7>
> .
>

rgommers commented:
This all looks good, comments about keeping the new functions private have been addressed. This was about to be merged 2 months ago, all we forgot to do was press the green button it looks like. So in it goes. Thanks @pvanmulbregt & all reviewers!
